### ISSUE TITLE
Dynamically generate entrypoints for JS assets

### ISSUE THREAD

**Issue by godber** [terascope/teraslice#2769]
I think we can do something like this to dynamically generate the asset registry for Teraslice assets:

```
'use strict';

const glob = require('glob');

glob('**/{processor,schema}.js', { ignore: ['node_modules/**', '__lib/**'] }, (er, files) => {
    console.log(files);
});
```

This would force the following requirements on these assets:

* All asset processor code be in `processor.js` or `schama.js` (I think this is a fixed, well known set of files, right? I haven't checked the loader code).
* All helper code would need to be in `__lib/` or be certain to not contain `.js` files that match those listed above

If the asset registry is ONLY used to supply esbuild entrypoints, then I think this should work.  Code like this could be added to the `teraslice-cli asset build` code to populate entrypoints.  If the registry is used in some other way, then perhaps this will be insufficient.

**Comment by peterdemartini** [comment]
I don't like the idea of this, mainly because this file won't exist until build time leading problems that won't be caught in CI. I think we'd be better off having a `teraslice-cli assets init --registry` command that would generate it for you. This way it will be ran as needed so the file will be checked-in and tested via the test-harness. Plus since it uses `yeoman` we get file-system diffing capabilities when the file already exists.

**Comment by peterdemartini** [comment]
As for some of the file conventions, ignore a directory or filename that starts with `__`, `.` or equals `node_modules`. And yes there is a strict file naming convention so requiring the files like that should be fine

**Comment by godber** [comment]
> I don't like the idea of this, mainly because this file won't exist until build time leading problems that won't be caught until CI publish step. I think we'd be better off having a `teraslice-cli assets init --registry` command that would generate it for you. This way it will be ran as needed so the file will be checked-in and tested via the test-harness. Plus since it uses `yeoman` we get file-system diffing capabilities when the file already exists.

That solution had not occurred to me and it seems like a good idea ... it adds friction to the asset development workflow though.  Would we expect that the generated `index.js` be committed to git?

**Comment by godber** [comment]
Is the registry used for anything other than providing an entrypoint for `esbuild`?

**Comment by godber** [comment]
BTW, this is motivated by two main things:

* avoiding writing boilerplate code
* (more importantly) help prevent mistakes where processors are accidentally omitted from a manually generate registry.

**Comment by peterdemartini** [comment]
>  ... it adds friction to the asset development workflow though.

Yes it would be a little bit more work for development but overall I think it gives the developer more control.

> Would we expect that the generated index.js be committed to git?

Yes the `index.js` or `index.ts` should be committed to git this way changes can be observed and tracked. 

> Is the registry used for anything other than providing an entrypoint for esbuild?

The registry _should_ be used in the tests, assuming the `WorkerTestHarness` or `SlicerTestHarness` is used.

**Comment by peterdemartini** [comment]
Also there is no reason we shouldn't be able to do the same thing for the typescript registry too
